In [12]:
import torch
import torchvision
import torchvision.transforms as transforms
from torch.utils.data import DataLoader

# Define transforms to preprocess the data
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])

# Download the MNIST datasets
train_dataset = torchvision.datasets.MNIST(root='./data', train=True, transform=transform, download=True)
test_dataset = torchvision.datasets.MNIST(root='./data', train=False, transform=transform, download=True)

# Create data loaders
batch_size = 100
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)


100%|██████████| 9912422/9912422 [00:00<00:00, 34319844.12it/s]


Extracting ./data/MNIST/raw/train-images-idx3-ubyte.gz to ./data/MNIST/raw



100%|██████████| 28881/28881 [00:00<00:00, 81903782.17it/s]


Extracting ./data/MNIST/raw/train-labels-idx1-ubyte.gz to ./data/MNIST/raw



100%|██████████| 1648877/1648877 [00:00<00:00, 64066878.47it/s]

Extracting ./data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./data/MNIST/raw


100%|██████████| 4542/4542 [00:00<00:00, 13559095.21it/s]


Extracting ./data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/MNIST/raw



In [7]:
import torch.nn as nn
import torch.nn.functional as F

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(28 * 28, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 10)

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

# Initialize the model
model = Net()


In [ ]:
import torch.optim as optim

# Define the loss function and optimizer
optimizer = optim.SGD(model.parameters(), lr=0.01)
criterion = nn.CrossEntropyLoss()

def train(model, train_loader, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}")

def test(model, test_loader):
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            inputs, labels = data
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    accuracy = 100 * correct / total
    return accuracy

# Train the model
train(model, train_loader, criterion, optimizer, num_epochs=25)


Epoch 1, Loss: 1.2929398519794146
Epoch 2, Loss: 0.465682396988074
Epoch 3, Loss: 0.37238926728566485
Epoch 4, Loss: 0.3326904684801896
Epoch 5, Loss: 0.3063545715312163
Epoch 6, Loss: 0.2851218855753541
Epoch 7, Loss: 0.2669487547377745
Epoch 8, Loss: 0.2504315092911323
Epoch 9, Loss: 0.23541991130759318
Epoch 10, Loss: 0.22135278141126036
Epoch 11, Loss: 0.2083909743403395
Epoch 12, Loss: 0.19651292011141777
Epoch 13, Loss: 0.18562815137207508
Epoch 14, Loss: 0.17598263083025814
Epoch 15, Loss: 0.16691079222907623
Epoch 16, Loss: 0.15913317969689766
Epoch 17, Loss: 0.15181762443855404
Epoch 18, Loss: 0.14463093611722191
Epoch 19, Loss: 0.13837042587498824
Epoch 20, Loss: 0.13277095414698123
Epoch 21, Loss: 0.12683287946507335
Epoch 22, Loss: 0.12232867147152622
Epoch 23, Loss: 0.1177519887375335
Epoch 24, Loss: 0.11299941732548177
Epoch 25, Loss: 0.10899771963556608


In [ ]:
final_accuracy = test(model, test_loader)
print(f"Final Validation Accuracy: {final_accuracy}%")


Final Validation Accuracy: 96.5%


In [3]:
!pip install aihwkit

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.3/11.3 MB 44.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 619.9/619.9 MB 2.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.4/294.4 kB 30.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.0/21.0 MB 70.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 849.3/849.3 kB 64.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.8/11.8 MB 97.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 557.1/557.1 MB 2.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.1/317.1 MB 3.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 168.4/168.4 MB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.6/54.6 MB 10.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 102.6/102.6 MB 8.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 173.2/173.2 MB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━

In [8]:
from aihwkit.nn import AnalogLinear
from aihwkit.optim import AnalogSGD
from aihwkit.simulator.configs import SingleRPUConfig
from aihwkit.simulator.configs.devices import ConstantStepDevice
import torch.optim as optim

# Define the custom neural network with AnalogLinear layers
class AnalogNet(nn.Module):
    def __init__(self):
        super(AnalogNet, self).__init__()
        self.config = SingleRPUConfig(device=ConstantStepDevice(w_min=-0.4))
        self.fc1 = AnalogLinear(28 * 28, 128)  # Replace Linear with AnalogLinear
        self.fc2 = AnalogLinear(128, 64)     # Replace Linear with AnalogLinear
        self.fc3 = AnalogLinear(64, 10)      # Replace Linear with AnalogLinear

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [11]:
# Initialize the model
model_const = AnalogNet()

# Apply regroup_param_groups to the model parameters (required by the tool)
analog_optimizer_const = AnalogSGD(model_const.parameters(),lr=0.01)
analog_optimizer_const.regroup_param_groups(model_const)

# Define the loss function
criterion = nn.CrossEntropyLoss()

def train(model, train_loader, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}")

# Train the model
train(model_const, train_loader, criterion, analog_optimizer_const, num_epochs=25)

Epoch 1, Loss: 1.3792122376461824
Epoch 2, Loss: 0.5321631866693497
Epoch 3, Loss: 0.45754598346849285
Epoch 4, Loss: 0.4273408761372169
Epoch 5, Loss: 0.42180571280419826
Epoch 6, Loss: 0.5050675835212072
Epoch 7, Loss: 0.615622464766105
Epoch 8, Loss: 0.6494255693505208
Epoch 9, Loss: 0.6324662974228462
Epoch 10, Loss: 0.6311758956561486
Epoch 11, Loss: 0.6379850436747074
Epoch 12, Loss: 0.6382508234431347
Epoch 13, Loss: 0.6468948687613011
Epoch 14, Loss: 0.6456993545095125
Epoch 15, Loss: 0.6525525434563557
Epoch 16, Loss: 0.6671155874431133
Epoch 17, Loss: 0.6657195872813463
Epoch 18, Loss: 0.6694677875687679
Epoch 19, Loss: 0.6741464104751745
Epoch 20, Loss: 0.669849790285031
Epoch 21, Loss: 0.6790625722954671
Epoch 22, Loss: 0.6810973875224591
Epoch 23, Loss: 0.6790054265906413
Epoch 24, Loss: 0.6874910006672144
Epoch 25, Loss: 0.6786023711164793


In [12]:
# Test the model and report the final accuracy
def test(model, test_loader):
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            inputs, labels = data
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    accuracy = 100 * correct / total
    return accuracy

final_accuracy_const = test(model_const, test_loader)
print(f"Final Validation Accuracy: {final_accuracy_const}%")

Final Validation Accuracy: 84.31%


In [51]:
help(ReRamESPresetDevice())

Help on ReRamESPresetDevice in module aihwkit.simulator.presets.devices object:

class ReRamESPresetDevice(aihwkit.simulator.configs.devices.ExpStepDevice)
 |  ReRamESPresetDevice(construction_seed: int = 0, corrupt_devices_prob: float = 0.0, corrupt_devices_range: float = 0.1, diffusion: float = 0.0, diffusion_dtod: float = 0.0, drift: aihwkit.simulator.parameters.utils.DriftParameter = <factory>, dw_min: float = 0.00135, dw_min_dtod: float = 0.2, dw_min_dtod_log_normal: bool = False, dw_min_std: float = 5.0, enforce_consistency: bool = True, lifetime: float = 0.0, lifetime_dtod: float = 0.0, perfect_bias: bool = False, reset: float = 0.0, reset_dtod: float = 0.0, reset_std: float = 0.01, up_down: float = 0.259359, up_down_dtod: float = 0.05, w_max: float = 1.0, w_max_dtod: float = 0.3, w_min: float = -1.0, w_min_dtod: float = 0.3, count_pulses: bool = False, A_up: float = -1.18445, A_down: float = -0.081404, gamma_up: float = 5.0, gamma_down: float = 5.0, a: float = -0.5, b: float = 

In [9]:
from aihwkit.simulator.presets import ReRamSBPresetDevice,ReRamESPresetDevice,CapacitorPresetDevice, IdealizedPresetDevice,PCMPresetDevice

In [42]:
class PresetNet_Re_SB(nn.Module):
    def __init__(self):
        super(PresetNet_Re_SB, self).__init__()
        config = SingleRPUConfig(device=ReRamSBPresetDevice(dw_min=0.0005))
        self.fc1 = AnalogLinear(28 * 28, 128,rpu_config=config)  # Replace Linear with AnalogLinear
        self.fc2 = AnalogLinear(128, 64, rpu_config=config)     # Replace Linear with AnalogLinear
        self.fc3 = AnalogLinear(64, 10, rpu_config=config)      # Replace Linear with AnalogLinear

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [43]:
# Initialize the model
model_re_sb = PresetNet_Re_SB()

# Apply regroup_param_groups to the model parameters (required by the tool)
analog_optimizer_re_sb = AnalogSGD(model_re_sb.parameters(),lr=0.01)
analog_optimizer_re_sb.regroup_param_groups(model_re_sb)

# Define the loss function
criterion = nn.CrossEntropyLoss()

def train(model, train_loader, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}")

# Train the model
train(model_re_sb, train_loader, criterion, analog_optimizer_re_sb, num_epochs=25)

Epoch 1, Loss: 2.112740519841512
Epoch 2, Loss: 1.348894037604332
Epoch 3, Loss: 1.0197316985328992
Epoch 4, Loss: 0.877666807572047
Epoch 5, Loss: 0.8270255957047145
Epoch 6, Loss: 0.7708141126235326
Epoch 7, Loss: 0.7512515611449877
Epoch 8, Loss: 0.7349214140574137
Epoch 9, Loss: 0.7176548382639885
Epoch 10, Loss: 0.701624760478735
Epoch 11, Loss: 0.6744586074848969
Epoch 12, Loss: 0.6706330042084058
Epoch 13, Loss: 0.667718369414409
Epoch 14, Loss: 0.6539174916346868
Epoch 15, Loss: 0.6585526502629121
Epoch 16, Loss: 0.6483133211731911
Epoch 17, Loss: 0.6359243289629618
Epoch 18, Loss: 0.6326238213479519
Epoch 19, Loss: 0.6326642408470313
Epoch 20, Loss: 0.6211604031423728
Epoch 21, Loss: 0.6160159929593404
Epoch 22, Loss: 0.610080925822258
Epoch 23, Loss: 0.6039068156977495
Epoch 24, Loss: 0.61463626096646
Epoch 25, Loss: 0.6096699408193429


In [50]:
# Test the model and report the final accuracy
def test(model, test_loader):
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            inputs, labels = data
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    accuracy = 100 * correct / total
    return accuracy

final_accuracy_re_sb = test(model_re_sb, test_loader)
print(f"Final Validation Accuracy: {final_accuracy_re_sb}%")

Final Validation Accuracy: 80.7%


In [14]:
class PresetNet_Re_ES(nn.Module):
    def __init__(self):
        super(PresetNet_Re_ES, self).__init__()
        config = SingleRPUConfig(device=ReRamESPresetDevice(dw_min=0.003))
        self.fc1 = AnalogLinear(28 * 28, 128,rpu_config=config)  # Replace Linear with AnalogLinear
        self.fc2 = AnalogLinear(128, 64, rpu_config=config)     # Replace Linear with AnalogLinear
        self.fc3 = AnalogLinear(64, 10, rpu_config=config)      # Replace Linear with AnalogLinear

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [15]:
# Initialize the model
model_re_es = PresetNet_Re_ES()

# Apply regroup_param_groups to the model parameters (required by the tool)
analog_optimizer_re_es = AnalogSGD(model_re_es.parameters(),lr=0.01)
analog_optimizer_re_es.regroup_param_groups(model_re_es)

# Define the loss function
criterion = nn.CrossEntropyLoss()

def train(model, train_loader, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}")

# Train the model
train(model_re_es, train_loader, criterion, analog_optimizer_re_es, num_epochs=25)

Epoch 1, Loss: 2.6786723081270853
Epoch 2, Loss: 2.313918716907501
Epoch 3, Loss: 2.3014424534638724
Epoch 4, Loss: 2.301323354244232
Epoch 5, Loss: 2.3103964948654174
Epoch 6, Loss: 2.3164948149522147
Epoch 7, Loss: 2.308799015680949
Epoch 8, Loss: 2.3012700502077736
Epoch 9, Loss: 2.301670427719752
Epoch 10, Loss: 2.3012713102499642
Epoch 11, Loss: 2.301224073966344
Epoch 12, Loss: 2.303142362833023
Epoch 13, Loss: 2.3142498970031737
Epoch 14, Loss: 2.3012070628007253
Epoch 15, Loss: 2.3011991421381635
Epoch 16, Loss: 2.3012215733528136
Epoch 17, Loss: 2.3011959783236184
Epoch 18, Loss: 2.3021786181132
Epoch 19, Loss: 2.3012188732624055
Epoch 20, Loss: 2.301214553117752
Epoch 21, Loss: 2.301218505303065
Epoch 22, Loss: 2.3011983931064606
Epoch 23, Loss: 2.3012008325258893
Epoch 24, Loss: 2.3012014512221017
Epoch 25, Loss: 2.301199638446172


In [16]:
# Test the model and report the final accuracy
def test(model, test_loader):
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            inputs, labels = data
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    accuracy = 100 * correct / total
    return accuracy

final_accuracy_re_es = test(model_re_es, test_loader)
print(f"Final Validation Accuracy: {final_accuracy_re_es}%")

Final Validation Accuracy: 11.35%


In [21]:
class PresetNet_CapPreDev(nn.Module):
    def __init__(self):
        super(PresetNet_CapPreDev, self).__init__()
        config = SingleRPUConfig(device=CapacitorPresetDevice())
        self.fc1 = AnalogLinear(28 * 28, 128,rpu_config=config)  # Replace Linear with AnalogLinear
        self.fc2 = AnalogLinear(128, 64, rpu_config=config)     # Replace Linear with AnalogLinear
        self.fc3 = AnalogLinear(64, 10, rpu_config=config)      # Replace Linear with AnalogLinear

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [22]:
# Initialize the model
model_CapPreDev = PresetNet_CapPreDev()

# Apply regroup_param_groups to the model parameters (required by the tool)
analog_optimizer_CapPreDev = AnalogSGD(model_CapPreDev.parameters(),lr=0.01)
analog_optimizer_CapPreDev.regroup_param_groups(model_CapPreDev)

# Define the loss function
criterion = nn.CrossEntropyLoss()

def train(model, train_loader, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}")

# Train the model
train(model_CapPreDev, train_loader, criterion, analog_optimizer_CapPreDev, num_epochs=25)

Epoch 1, Loss: 1.3855879114568233
Epoch 2, Loss: 0.6058886830012004
Epoch 3, Loss: 0.6914633496602376
Epoch 4, Loss: 0.7654592626790205
Epoch 5, Loss: 0.7929244703551134
Epoch 6, Loss: 0.8186947929362456
Epoch 7, Loss: 0.8375282887617747
Epoch 8, Loss: 0.8489544479052226
Epoch 9, Loss: 0.8784996323784192
Epoch 10, Loss: 0.8783290661374727
Epoch 11, Loss: 0.8870744270086288
Epoch 12, Loss: 0.8948428469896317
Epoch 13, Loss: 0.8987029077112675
Epoch 14, Loss: 0.9032369999587536
Epoch 15, Loss: 0.9049132996797562
Epoch 16, Loss: 0.9091054199139277
Epoch 17, Loss: 0.9108297832806905
Epoch 18, Loss: 0.9139126645276944
Epoch 19, Loss: 0.9229602523644765
Epoch 20, Loss: 0.9064821149408817
Epoch 21, Loss: 0.8986120115717252
Epoch 22, Loss: 0.9081906072298686
Epoch 23, Loss: 0.9271186724801859
Epoch 24, Loss: 0.9291483777264754
Epoch 25, Loss: 0.9312358812491099


In [23]:
# Test the model and report the final accuracy
def test(model, test_loader):
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            inputs, labels = data
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    accuracy = 100 * correct / total
    return accuracy

final_accuracy_CapPreDev = test(model_CapPreDev, test_loader)
print(f"Final Validation Accuracy: {final_accuracy_CapPreDev}%")

Final Validation Accuracy: 81.01%


In [24]:
class PresetNet_IdPreDev(nn.Module):
    def __init__(self):
        super(PresetNet_IdPreDev, self).__init__()
        config = SingleRPUConfig(device=IdealizedPresetDevice())
        self.fc1 = AnalogLinear(28 * 28, 128,rpu_config=config)  # Replace Linear with AnalogLinear
        self.fc2 = AnalogLinear(128, 64, rpu_config=config)     # Replace Linear with AnalogLinear
        self.fc3 = AnalogLinear(64, 10, rpu_config=config)      # Replace Linear with AnalogLinear

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [25]:
# Initialize the model
model_IdPreDev = PresetNet_IdPreDev()

# Apply regroup_param_groups to the model parameters (required by the tool)
analog_optimizer_IdPreDev = AnalogSGD(model_IdPreDev.parameters(),lr=0.01)
analog_optimizer_IdPreDev.regroup_param_groups(model_IdPreDev)

# Define the loss function
criterion = nn.CrossEntropyLoss()

def train(model, train_loader, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}")

# Train the model
train(model_IdPreDev, train_loader, criterion, analog_optimizer_IdPreDev, num_epochs=25)

Epoch 1, Loss: 1.380493186612924
Epoch 2, Loss: 0.48755881217618785
Epoch 3, Loss: 0.3924475624660651
Epoch 4, Loss: 0.3557643955945969
Epoch 5, Loss: 0.33271531594296294
Epoch 6, Loss: 0.3114937826246023
Epoch 7, Loss: 0.2939030200615525
Epoch 8, Loss: 0.2767156941567858
Epoch 9, Loss: 0.26670783532162506
Epoch 10, Loss: 0.25967533137649296
Epoch 11, Loss: 0.24348721948141852
Epoch 12, Loss: 0.23503106214106084
Epoch 13, Loss: 0.22457356059302885
Epoch 14, Loss: 0.21292693580811223
Epoch 15, Loss: 0.20391046931346257
Epoch 16, Loss: 0.19272308034201463
Epoch 17, Loss: 0.18534941354145607
Epoch 18, Loss: 0.18033011083801587
Epoch 19, Loss: 0.17540942346677185
Epoch 20, Loss: 0.16583890946581958
Epoch 21, Loss: 0.16085710107038417
Epoch 22, Loss: 0.15526536073846123
Epoch 23, Loss: 0.151482256182159
Epoch 24, Loss: 0.14511149413262805
Epoch 25, Loss: 0.13906903809557358


In [26]:
# Test the model and report the final accuracy
def test(model, test_loader):
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            inputs, labels = data
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    accuracy = 100 * correct / total
    return accuracy

final_accuracy_IdPreDev = test(model_IdPreDev, test_loader)
print(f"Final Validation Accuracy: {final_accuracy_IdPreDev}%")

Final Validation Accuracy: 95.48%


In [27]:
class PresetNet_PCMPreDev(nn.Module):
    def __init__(self):
        super(PresetNet_PCMPreDev, self).__init__()
        config = SingleRPUConfig(device=PCMPresetDevice())
        self.fc1 = AnalogLinear(28 * 28, 128,rpu_config=config)  # Replace Linear with AnalogLinear
        self.fc2 = AnalogLinear(128, 64, rpu_config=config)     # Replace Linear with AnalogLinear
        self.fc3 = AnalogLinear(64, 10, rpu_config=config)      # Replace Linear with AnalogLinear

    def forward(self, x):
        x = x.view(-1, 28 * 28)  # Flatten the input
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [28]:
# Initialize the model
model_PCMPreDev = PresetNet_PCMPreDev()

# Apply regroup_param_groups to the model parameters (required by the tool)
analog_optimizer_PCMPreDev = AnalogSGD(model_PCMPreDev.parameters(),lr=0.01)
analog_optimizer_PCMPreDev.regroup_param_groups(model_PCMPreDev)

# Define the loss function
criterion = nn.CrossEntropyLoss()

def train(model, train_loader, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch + 1}, Loss: {running_loss / len(train_loader)}")

# Train the model
train(model_PCMPreDev, train_loader, criterion, analog_optimizer_PCMPreDev, num_epochs=25)

Epoch 1, Loss: 2.304488911231359
Epoch 2, Loss: 2.3019596695899964
Epoch 3, Loss: 2.301759931643804
Epoch 4, Loss: 2.3014686632156374
Epoch 5, Loss: 2.3014335946242013
Epoch 6, Loss: 2.301485204696655
Epoch 7, Loss: 2.3016721030076344
Epoch 8, Loss: 2.3014841349919637
Epoch 9, Loss: 2.3013617797692616
Epoch 10, Loss: 2.3012955558300017
Epoch 11, Loss: 2.301239277124405
Epoch 12, Loss: 2.3012940259774526
Epoch 13, Loss: 2.301286664009094
Epoch 14, Loss: 2.3012045005957287
Epoch 15, Loss: 2.301221194267273
Epoch 16, Loss: 2.301252621014913
Epoch 17, Loss: 2.3012128098805746
Epoch 18, Loss: 2.3012103907267254
Epoch 19, Loss: 2.3012085322539013
Epoch 20, Loss: 2.301207369963328
Epoch 21, Loss: 2.3012103895346323
Epoch 22, Loss: 2.301208261648814
Epoch 23, Loss: 2.301203666130702
Epoch 24, Loss: 2.3012048637866975
Epoch 25, Loss: 2.30120094537735


In [29]:
# Test the model and report the final accuracy
def test(model, test_loader):
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            inputs, labels = data
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    accuracy = 100 * correct / total
    return accuracy

final_accuracy_PCMPreDev = test(model_PCMPreDev, test_loader)
print(f"Final Validation Accuracy: {final_accuracy_PCMPreDev}%")

Final Validation Accuracy: 11.35%


## Summary Table
| Device/Model | Accuracy |
|-----------------|-----------------|
| Pytorch   | 96.5%   |
| ConstantStepDevice   | 84.31%   |
| ReRamSBPresetDevice  | 80.70%   |
| ReRamESPresetDevice   | 11.35%   |
| CapacitorPresetDevice   | 81.01%  |
| IdealizedPresetDevice   | 95.48%  |
| PCMPresetDevice   | 11.35%  |